pdfminer/pdfdocument.py

```diff
@@ -500,28 +500,74 @@ def get_cfm(self, name):
 
     def authenticate(self, password):
         password = password.encode('utf-8')[:127]
-        hash = sha256(password)
-        hash.update(self.o_validation_salt)
-        hash.update(self.u)
-        if hash.digest() == self.o_hash:
-            hash = sha256(password)
-            hash.update(self.o_key_salt)
-            hash.update(self.u)
-            cipher = Cipher(algorithms.AES(hash.digest()),
+        hash = self._password_hash(password, self.o_validation_salt, self.u)
+        if hash == self.o_hash:
+            hash = self._password_hash(password, self.o_key_salt, self.u)
+            cipher = Cipher(algorithms.AES(hash),
                             modes.CBC(b'\0' * 16),
                             backend=default_backend())
             return cipher.decryptor().update(self.oe)
-        hash = sha256(password)
-        hash.update(self.u_validation_salt)
-        if hash.digest() == self.u_hash:
-            hash = sha256(password)
-            hash.update(self.u_key_salt)
-            cipher = Cipher(algorithms.AES(hash.digest()),
+        hash = self._password_hash(password, self.u_validation_salt)
+        if hash == self.u_hash:
+            hash = self._password_hash(password, self.u_key_salt)
+            cipher = Cipher(algorithms.AES(hash),
                             modes.CBC(b'\0' * 16),
                             backend=default_backend())
             return cipher.decryptor().update(self.ue)
         return None
 
+    def _password_hash(self, password, salt, vector=b""):
+        """
+        Compute password hash depending on revision number
+        """
+        if self.r == 5:
+            return self._r5_password(password, salt, vector)
+        return self._r6_password(password, salt[0:8], vector)
+
+    def _r5_password(self, password, salt, vector):
+        """
+        Compute the password for revision 5
+        """
+        hash = sha256(password)
+        hash.update(salt)
+        if vector:
+            hash.update(vector)
+        return hash.digest()
+
+    def _r6_password(self, password, salt, vector):
+        """
+        Compute the password for revision 6
+        """
```

Comment:
As I stated higher up in the thread, this is essentially a reformatted version of https://github.com/MatthiasValvekens/pyHanko/blob/b066f22bf35b60be9a8f3d8c0fe01d22850671bc/pyhanko/pdf_utils/crypt.py#L234-L263 (restating it here to make the reference more precise).

I'd appreciate a license note in accordance with the terms of the MIT License, thanks :)

Reply:
I added the license and an acknowledgment section in the README